Support BreakingGround-DLC in instance faking

DasSkelett opened the pull request:
Got a bit bigger than I thought.
(And it's getting bigger and bigger)

Changes
-----
Important stuff first:
Console command for faking changed!
It is now `ckan ksp fake <name> <path> <version> [--MakingHistory <version>] [--BreakingGround <version>]`
instead of previous `ckan ksp fake <name> <path> <version> [MakingHistoryVersion]`.
I think this is more future-proof.
I am going to try my luck adjusting KSP-CKAN/xKAN-meta_testing.


**`IDlcDetector`**:
I added `string InstallPath()`, it returns the install path relative to the GameDir, that means for example `"GameData/SquadExpansion/Serenity"`.
Depending on how [this idea of mine](https://github.com/KSP-CKAN/CKAN/issues/2770#issuecomment-498413897) will be implemented (if at all), we might restructure the whole thing again. But I think this is the best solution for now, I'm open for feedback.

I added `bool AllowedOnBaseVersion(KspVersion)`, it returns whether a DLC is allowed to be installed or faked on a given basegame version.

I added
```csharp
string IdentifierBaseName { get; }
KspVersion ReleaseGameVersion { get; }
```
to allow accessing those two from outside the interface implementations.
I changed the two fields with these names inside `StandardDlcDetectorBase` to getter without setter (this effectively is like the `readonly` modificator before).

**`CloneFakeInstanceDialog`**:
The GUI `CloneFakeInstanceDialog` has been adjusted:
![CloneFakeInstanceDialog](https://user-images.githubusercontent.com/28812678/59028367-c4ffa300-885b-11e9-84cf-6bcaa0442e4e.png)
If Squad releases another 10 DLCs we might get out of space and need a new system, but for now it works.

**`FakeInstance()`**:
`FakeInstance()` is now a transaction.
Before the DLC just wasn't faked if f.e. the KSP version was < 1.4.0.
This left the user in an unfavorable situation, he had to delete the new folder and run `ckan ksp forget` once and try again to include the DLC. now it's all or nothing.

`FakeInstance()` takes the DLCs that should be faked as `Dictionary<IDlcDetector, KspVersion>`. Future DLC releases _shouldn't_ need a change in this fuction.

Also `CopyDirectory()` is now a real transaction that includes IO operations, if I understand stock .NET transactions right they don't include them, but using `ChinhDo.Transactions` with `TxFileManager` should fix this.
Correct me if I got something wrong there (or anywhere else too of course).

Fixes #2791 

~Oh yeah, #2749 needs another rebase ^^~

DasSkelett commented on Core/Utilities.cs:
```diff
@@ -1,6 +1,7 @@
-﻿using System;
+using System;
```
Here was a BOM (0xFEFF) at the beginning, I removed it.

DasSkelett commented:
I see that the fact that DLCs weren't faked without error message if the version was too low was used in the validation scripts.
But I still believe that we should do all or nothing.

We can use the `ckan compare` command to determine whether to fake a DLC or not in the scripts.
Right now it outputs a very human readable answer:
`"1.1.0" is lower than "1.4.1".`
but if we add another option (`--machine-readable` or so) to output -1, 0 or 1 it should be more usable.

HebaruSan commented:
The scripts have a function `versions_less_or_equal` to do version comparisons.

DasSkelett commented:
> The scripts have a function `versions_less_or_equal` to do version comparisons.

That's cool, thanks!

HebaruSan commented:
> Oh yeah, #2749 needs another rebase ^^

The shoe is on the other foot!

DasSkelett commented:
> The shoe is on the other foot!

Darn it, feared this would happen.
Will take some days until I'm home again.

HebaruSan commented:
Created DasSkelett/CKAN#5. GitHub reports conflicts, which shouldn't surprise it since rebasing was the point.

DasSkelett commented:
'Merged' the PR, thanks.

HebaruSan commented on Cmdline/Action/KSP.cs:
```diff
@@ -550,36 +564,48 @@ private int FakeNewKSPInstall (FakeOptions options)
-                Manager.FakeInstance(installName, path, version, dlcVersion);
-                if (setToDefault)
+                Manager.FakeInstance(installName, path, version, dlcVersions);
+                if (setDefault)
```
We are missing curly braces here. Only the `RaiseMessage` call is controlled by the `if`, the `SetAutoStart` will happen unconditionally.

DasSkelett replied:
A bit of Python here and there can't hurt.

... actually, it does in this case.

HebaruSan commented on Cmdline/Action/KSP.cs:
```diff
@@ -124,8 +123,13 @@ internal class FakeOptions : CommonOptions
+
+            [Option("MakingHistory", DefaultValue = "none", HelpText = "The version of the DLC to be faked.")]
+            public string makingHistoryVersion { get; set; }
+            [Option("BreakingGround", DefaultValue = "none", HelpText = "The version of the DLC to be faked.")]
```
Let's make the help texts different. "... of the Making History DLC" ?

HebaruSan commented on Core/KSPManager.cs:
```diff
@@ -200,67 +202,96 @@ public void CloneInstance (KSP existing_instance, string new_name, string new_pa
         /// <param name="version">The version of the new instance. Should have a build number.</param>
-        /// <param name="dlcVersion">The version of the DLC. Null if DLC should be faked.</param>
-        public void FakeInstance(string new_name, string new_path, KspVersion version, string dlcVersion = null)
+        /// <param name="dlcVersions">The versions of the DLCs. Null / empty strings if DLCs should not be faked.</param>
```
I'm not a big fan of an array of strings as a parameter type, with the array indices meaningful and `null` padding values required in some cases. It'll require us to update this function every time a new DLC is released, and it's prone to array out of bounds errors and so forth. And the current parameter documentation doesn't explain how to use it.

What do you think about changing this parameter to a `Dictionary<IDlcDetector, KspVersion>` ? Then we could `foreach` through them to get the key value pairs, and use `InstallPath()` on the key to create the folders in a generic way. If you want to retain the game version validation, we could add a `bool AllowedOnBaseVersion(KspVersion)` method to `IDlcDetector` to do that generically as well. Then when a new DLC is released we would only have to update the calling code to add a new pair to the dictionary. Generating that parameter should be relatively simple:

```csharp
    new Dictionary<IDlcDetector, KspVersion>() {
        { new MakingHistoryDlcDetector(),  new KspVersion("1.1.0") },
        { new BreakingGroundDlcDetector(), new KspVersion("1.0.0") },
    }
```

DasSkelett replied:
Sounds like a much better solution than mine. 👍

DasSkelett replied:
I would add a new `private readonly KspVersion ReleaseGameVersion;` to the `StandardDlcDetectorBase` to save the KSP version there, and set it in the constructors like `IdentifierBaseName` and `DirectoryBaseName`.
I think it makes sense to have it in an obvious, code-accessible place, and not only buried somewhere inside `bool AllowedOnBaseVersion()`.
Does that seem reasonable to you?

HebaruSan replied:
Definitely make whatever changes you think are appropriate for the `AllowedOnBaseVersion` concept, that was just meant to be very vague pseudocode. :+1:

DasSkelett replied:
I'll try my best 😆

DasSkelett replied:
One more thing:
Right now we output the DLC version as an `UnmanagedModuleVersion` in the detectors.
Should we 'upgrade' that to `KspVersion`? It allows better handling, and if I'm not mistaken it always matched the pattern, and probably will do so in the future too.

HebaruSan replied:
I guess that sounds fine to me, but let's ask @dbent in case there are subtleties that I'm not noticing.

dbent replied:
The version object type is used as a sentinel value in a few different places to distinguish between modules CKAN installed itself and those that were autodetected. I don't particularly like that strategy myself but it's a legacy from how the original autodetect code was written.

If you return a `ModuleVersion` instead of an `UnmanagedModuleVersion` then everywhere that checks if the version object is an `UnmanagedModuleVersion` will have to be updated. IIRC, the changes were numerous and sensitive enough that it wasn't worth the effort at the time.

DasSkelett replied:
Okay, thanks @dbent. So not something I want to do in this PR, if at all ^^
If this PR stays as is, we would be double-tracked, with all the instance faking stuff using `KspVersion` and the mod managing and other stuff using `UnmanagedModuleVersion`.
Well, it works, but I don't make use of it being `KspVersion`, so I probably should change it (back) to `UnmanagedModuleVersion` for the sake of consistency.

DasSkelett replied:
On another thought, I could just leave it as is. The format checks should be done either way, and it should simplify a future change to `KspVersion` in the underlying module managing code.

HebaruSan replied:
The new DLC creation implementation looks good now. :+1:

DasSkelett commented on Core/KSP.cs:
```diff
@@ -440,6 +445,7 @@ public bool ScanGameData()
 
                 bool dlcChanged = manager.ScanDlc();
                 manager.Save(false);
+                manager.Dispose();
```
@HebaruSan, is this safe to call here?
Problem was, `SetupCkanDirectories()` calls `ScanGameData()`,
which creates a new `RegistryManager.Instance()`, which creates a lock file.
This interferes with transactions.

Since this `RegistryManager` object is instantiated inside `ScanGameData()`, it can't be accessed after leaving the function scope, yes? This should make it safe to dispose it.

I tried to add a new KSP dir without a dir, which calls `ScanGameData()` too, and that still works fine :D

HebaruSan replied:
> Since this RegistryManager object is instantiated inside ScanGameData(), it can't be accessed after leaving the function scope, yes?

No, any other code can access the same manager via `RegistryManager.Instance(ksp)` if it has a reference to the same `KSP ksp` object. In fact, it's possible that some other code *already* has such a reference and is depending on it not being disposed (GUI calls this function at the start of `UpdateRepo` and probably would prefer to keep its lock).

DasSkelett replied:
Oh well, I see it's a bad idea.
Then I'll search another way to remove the lock file in failed transactions or prevent it from being created at all.

DasSkelett replied:
I solved it with passing a bool to `KSP()` and `SetupCkanDirectories()` signaling whether `ScanGameData()` should be called. It's the easiest way I think.

HebaruSan commented on Core/DLC/IDlcDetector.cs:
```diff
@@ -7,6 +7,9 @@ namespace CKAN.DLC
     public interface IDlcDetector
     {
+        string IdentifierBaseName { get; }
+        KspVersion ReleaseGameVersion { get; }
```
Please explicitly mark these as `public` or `private`.

HebaruSan commented:
(Please ignore my deleted comment about marking things public or private. I forgot I was looking at an Interface.)

HebaruSan commented on Core/KSPManager.cs:
```diff
@@ -200,67 +206,91 @@ public void CloneInstance (KSP existing_instance, string new_name, string new_pa
+                }
+                // Thrown by AddInstance() if created instance is not valid.
+                // Thrown f.e. if a write operation didn't complete for unknown reasons.
+                catch (NotKSPDirKraken kraken)
```
Unless I'm missing something, we don't need to catch and re-throw here. That's for when you want to throw a different exception or take some action in addition. If we just remove the try/catch blocks, the effect should be the same.

DasSkelett replied:
Yeah, I added that for clarity, to point out what call can throw which exception and why. Don't know if that's a 'don't' in C#.

HebaruSan replied:
Oh, OK. Nevermind then!

HebaruSan commented on Core/KSP.cs:
```diff
@@ -71,24 +72,29 @@ public KSP(string gameDir, string name, IUser user)
-                Directory.CreateDirectory(CkanDir());
+                txFileMgr.CreateDirectory(CkanDir());
 
                 User.RaiseMessage("Scanning for installed mods...");
```
Very minor, but we probably don't need this `RaiseMessage` if `scan` is false.

HebaruSan commented:
Another minor point, the name error ends with "--Done--" while the folder error ends with "--Error--". Looks like this is because `Manager.HasInstance(installName)` will return true thanks to the instance that's already using the name?

```
$ _build/ckan.exe ksp fake Dakar /tmp/invalid2 1.5.1 --MakingHistory 1.0.0 --BreakingGround none
Creating new fake KSP install Dakar at /tmp/invalid2 with version 1.5.1.2335
This instance name is already taken: Dakar
--Done--
$ _build/ckan.exe ksp fake Dakar2 /tmp/invalid 1.5.1 --MakingHistory 1.0.0 --BreakingGround none
Creating new fake KSP install Dakar2 at /tmp/invalid with version 1.5.1.2335
The specified folder already exists and is not empty.
--Error--
```

DasSkelett commented:
> Another minor point, the name error ends with "--Done--" while the folder error ends with "--Error--". Looks like this is because `Manager.HasInstance(installName)` will return true thanks to the instance that's already using the name?

It's because I just forgot to put a `return` in that specific kraken catch, so the execution continues and yes, `Manager.HasInstance()` finally returns `true`.

DasSkelett commented on Tests/Core/KSPManager.cs:
```diff
@@ -195,20 +196,29 @@ public void FakeInstance_InNotEmptyFolder_ThrowsBadInstallLocationKraken()
         }
 
         [Test]
-        public void FakeInstance_ValidArgumentsWithDLC_ManagerHasValidInstance()
```
Is that new test as you imagined?

HebaruSan replied:
Yes, looks good, thanks!

DasSkelett commented:
Okay, some more changes:
* Added new kraken: `WrongKspVersionKraken`, to be used if a given KSP version is valid, but not allowed for an action (like instance faking and older base game version).
* Renamed `IncorrectKspVersionKraken` to `BadKspVersionKraken` to better distinct it from the new one. To be used if a version is not valid, i.e. not in the build map or malformatted.
* Added another test to check if `FakeInstance()` throws `WrongKspVersionkraken` for base game versions that are too old for the DLC(s) requested to fake.
* Redid the exit code handling in `Cmdline.KSP.FakeNewInstance()` a bit.

DasSkelett commented:
Does the bot skip network-dependent tests?

HebaruSan commented:
I believe so:

https://github.com/KSP-CKAN/CKAN/blob/1ac6261f0f81a07213c91c967c444ed451324990/.travis.yml#L43

DasSkelett commented:
Yep, just discovered that flag. Was wondering, because on my machine I obviously fail the two Curseforge tests.
